javascore/irc2Tradeable/src/main/java/foundation/icon/btp/irc2Tradeable/IRC2Basic.java

```diff
@@ -0,0 +1,144 @@
+    private final VarDB<String> symbol = Context.newVarDB("token_symbol", String.class);
+    private final VarDB<BigInteger> decimals = Context.newVarDB("decimals", BigInteger.class);
+    private final VarDB<BigInteger> totalSupply = Context.newVarDB("total_supply", BigInteger.class);
+    private final DictDB<Address, BigInteger> balances = Context.newDictDB("balances", BigInteger.class);
+
+    public IRC2Basic(String _name, String _symbol, int _decimals) {
+        // initialize values only at first deployment
+        if (this.name.get() == null) {
+            this.name.set(ensureNotEmpty(_name));
+            this.symbol.set(ensureNotEmpty(_symbol));
+
+            // decimals must be larger than 0 and less than 21
+            Context.require(_decimals >= 0, "decimals needs to be positive");
+            Context.require(_decimals <= 21, "decimals needs to be equal or lower than 21");
+            this.decimals.set(BigInteger.valueOf(_decimals));
+        }
+    }
+
+    private String ensureNotEmpty(String str) {
+        Context.require(str != null && !str.trim().isEmpty(), "str is null or empty");
+        assert str != null;
+        return str.trim();
+    }
+
+    @External(readonly=true)
+    public String name() {
+        return name.get();
+    }
+
+    @External(readonly=true)
+    public String symbol() {
+        return symbol.get();
+    }
+
+    @External(readonly=true)
+    public BigInteger decimals() {
+        return decimals.get();
+    }
+
+    @External(readonly=true)
+    public BigInteger totalSupply() {
+        return totalSupply.getOrDefault(BigInteger.ZERO);
+    }
+
+    @External(readonly=true)
+    public BigInteger balanceOf(Address _owner) {
+        return safeGetBalance(_owner);
+    }
+
+    @External
+    public void transfer(Address _to, BigInteger _value, @Optional byte[] _data) {
+        Address _from = Context.getCaller();
+        _transfer(_from, _to, _value, _data);
+    }
+
+    protected void _transfer(Address _from, Address _to, BigInteger _value, @Optional byte[] _data) {
+        Context.require(!ZERO_ADDRESS.equals(_from), "_from address cannot be zero address");
+        Context.require(!ZERO_ADDRESS.equals(_to), "_to address cannot be zero address");
+
+        // check some basic requirements
+        Context.require(_value.compareTo(BigInteger.ZERO) >= 0, "_value needs to be positive");
+        Context.require(safeGetBalance(_from).compareTo(_value) >= 0, "Insufficient balance");
+
+        // adjust the balances
+        safeSetBalance(_from, safeGetBalance(_from).subtract(_value));
+        safeSetBalance(_to, safeGetBalance(_to).add(_value));
+
+        // if the recipient is SCORE, call 'tokenFallback' to handle further operation
+        byte[] dataBytes = (_data == null) ? new byte[0] : _data;
+        if (_to.isContract()) {
+            Context.call(_to, "tokenFallback", _from, _value, dataBytes);
```

Comment:
Just making sure, the BSH contract still needs the tokenFallBack to maintain the balances, and the BSH has not been changed to get the balances from the IRC2 contract, if I'm not wrong?

Reply:
not sure i can understand